srml/staking/src/lib.rs

```diff
@@ -745,17 +754,30 @@ decl_module! {
 		/// See also [`Call::unbond`].
 		///
 		/// # <weight>
-		/// - Could be dependent on the `origin` argument and how much `unlocking` chunks exist. It implies
-		///   `consolidate_unlocked` which loops over `Ledger.unlocking`, which is indirectly
-		///   user-controlled. See [`unbond`] for more detail.
+		/// - Could be dependent on the `origin` argument and how much `unlocking` chunks exist.
+		///  It implies `consolidate_unlocked` which loops over `Ledger.unlocking`, which is
+		///  indirectly user-controlled. See [`unbond`] for more detail.
 		/// - Contains a limited number of reads, yet the size of which could be large based on `ledger`.
 		/// - Writes are limited to the `origin` account key.
 		/// # </weight>
 		fn withdraw_unbonded(origin) {
 			let controller = ensure_signed(origin)?;
 			let ledger = Self::ledger(&controller).ok_or("not a controller")?;
 			let ledger = ledger.consolidate_unlocked(Self::current_era());
-			Self::update_ledger(&controller, &ledger);
+
+			if ledger.unlocking.is_empty() && ledger.active.is_zero() {
```

Comment:
Not sure what this is for. If the (active) balance ever falls below ED, then these consequences will happen automatically.

Reply:
I think they are separate: an account can have 100 units and bond 50 of them while ED is 10, meaning that regardless of how/when they unbond, their total balance will remain above ED and won't be cleaned by the `OnFreeBalanceZero`. But, we still keep all of their relevant info in the staking module even if they `unbond` and `withdraw_unbonded` all of it out. This check essentially treats this situation.

Reply:
```
/// NOTE: Two of the storage writes (`Self::bonded`, `Self::payee`) are _never_ cleaned unless
/// the `origin` falls below _existential deposit_ and gets removed as dust.
```

I think forcing minimum ED for the `bond()` (almost) solves ^^ but what I added (cleaning after withdraw) is a further improvement on top of it to _not store the information of anyone who ever bonded and withdrawn_ __but has some balance in their account left__. if theu have unbonded and withdrawn, I think it is clear that those accounts might still live (for long) but have nothing more to do with the Staking system.